### PR TITLE
TDR-2450 - Add multi select drop down for closure metadata

### PR DESCRIPTION
**Pull request by vimleshtna**

### CODE REVIEW

File: app/controllers/util/DynamicFormUtils.scala [hunk 1]
```diff
@@ -11,6 +11,7 @@ class DynamicFormUtils(request: Request[AnyContent], defaultFieldValues: List[Fo
   lazy val formAnswersWithValidInputNames: Map[String, Seq[String]] = formAnswers.filter {
     case (inputName, _) if inputName.startsWith("input") => true
     case (inputName, _) if inputName == "csrfToken"      => false
+    case (inputName, _) if inputName.startsWith("tna-")  => false
```
Comment (TomJKing) [inline app/controllers/util/DynamicFormUtils.scala:14]: Is there a reason this input is being excluded? Does the name need to be more specific?

File: npm/src/index.ts [hunk 2]
```diff
@@ -42,6 +43,7 @@ export const renderModules = async () => {
   const fileNavigation = document.querySelector(".govuk-tna-tree")
   const timeoutDialog: HTMLDialogElement | null =
     document.querySelector(".timeout-dialog")
+  const multiSelectSearch = document.querySelector(".tna-multi-select-search")
```
Comment (TomJKing) [inline npm/src/index.ts:46]: Should update the unit tests to make sure the new component is called correctly: `/npm/test/index.test.ts`

**Review by TomJKing — COMMENTED** [review]
Looks good. Some minor comments / suggestions.

File: app/views/standard/addAdditionalMetadata.scala.html [hunk 3]
```diff
@@ -72,13 +73,13 @@ <h1 class="govuk-heading-l">Add or edit metadata</h1>
                   inputMode,
                   addSuffixText
                 )
-                case DropdownField(fieldId, fieldName, fieldDescription, multiValue, options, selectedOption, _, fieldErrors) =>
-                  inputDropdown(
+                case DropdownField(fieldId, fieldName, fieldDescription, multiValue, options, selectedOptions, _, fieldErrors) =>
```
[the remark this answers is not in the document]
Comment (vimleshtna) [reply app/views/standard/addAdditionalMetadata.scala.html]: Yeah, created a new one `MultiSelectField`.

File: app/controllers/util/FormField.scala [hunk 4]
```diff
@@ -122,18 +133,33 @@ object RadioButtonGroupField {
   }
 }
 
+object MultiSelectField {
+
+  def validate(slectedOptions: Seq[String], multiSelectField: MultiSelectField): Option[String] =
+    slectedOptions match {
```
Comment (TomJKing) [inline app/controllers/util/FormField.scala:139]: Spelling of parameter. Should be `selectedOptions`

File: app/controllers/util/CustomMetadataUtils.scala [hunk 5]
```diff
@@ -72,15 +72,27 @@ class CustomMetadataUtils(allCustomMetadataProperties: List[CustomMetadata]) {
...
-              property.defaultValue.map(value => InputNameAndValue(value, value)),
-              isRequired
-            )
+            if (property.multiValue) {
+              MultiSelectField(
+                property.name,
+                fieldLabel,
+                fieldDescription,
+                property.multiValue,
+                property.values.sortBy(_.uiOrdinal).map(v => InputNameAndValue(v.value, v.value)),
```
Comment (TomJKing) [inline app/controllers/util/CustomMetadataUtils.scala:81]: This value is used for either option, so maybe abstract out to a val? `val options = property.values.sortBy(_.uiOrdinal).map(v => InputNameAndValue(v.value, v.value))`

File: app/controllers/util/FormField.scala [hunk 6]
```diff
@@ -122,18 +133,33 @@ object RadioButtonGroupField {
...
+    selectedOption match {
+      case None                                                           => Some(dropdownOptionNotSelectedError.format(dropdownField.fieldName))
+      case Some(value) if !dropdownField.options.exists(_.value == value) => Some(invalidDropdownOptionSelectedError.format(value))
+      case _                                                              => None
     }
 
-  def update(dropdownField: DropdownField, value: String): DropdownField = {
-    val optionSelected: Option[InputNameAndValue] = if (value.isEmpty) None else Some(InputNameAndValue(value, value))
-    dropdownField.copy(selectedOption = optionSelected)
+  def update(multiSelectField: DropdownField, selectedValue: Option[String]): DropdownField = {
```
Comment (TomJKing) [inline app/controllers/util/FormField.scala:160]: Think the `multiSelectField` parameter should still be named: `dropdownField`, as this is a method for the `DropdownField` object.

File: app/views/partials/inputMultiSelect.scala.html [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+fieldErrors: List[String] = Nil
+)(implicit messages: Messages)
+
+@import viewsapi.FormFunctions._
+@import views.html.partials.errorMessage
+
+@disabledStatus = {@attributes.get(Symbol("_disabledOption"))}
+@isDynamicForm = @{attributes.getOrElse(Symbol("_dynamicForm"), false).asInstanceOf[Boolean]}
+@formFieldId = @{if(isDynamicForm) s"inputmultiselect-$fieldId" else fieldId}
+
```
Comment (TomJKing) [inline app/views/partials/inputMultiSelect.scala.html:18]: Unnecessary line break

File: npm/test/index.test.ts [hunk 8]
```diff
@@ -116,3 +116,35 @@ test("renderModules does not call authorisation when dialog box is not present o
 
   keycloakInstance.mockRestore()
 })
+
+test("renderModules should initialize the tna-multi-select-search element on the page", async () => {
```
Comment (TomJKing) [inline npm/test/index.test.ts:120]: Maybe reword the test slightly: `renderModules should initialise the multi-select search module if the tna-multi-select-search element is present on the page`

File: npm/test/index.test.ts [hunk 9]
```diff
@@ -116,3 +116,35 @@ test("renderModules does not call authorisation when dialog box is not present o
...
+  const keycloakInstance = getKeycloakInstance as jest.Mock
+  keycloakInstance.mockImplementation(() => Promise.resolve(mockKeycloak))
+
+  document.body.innerHTML =
+    `
+    <div class="tna-multi-select-search" data-module="multi-select-search">
+        <div class="tna-multi-select-search" data-module="multi-select-search">
```
Comment (TomJKing) [inline npm/test/index.test.ts:127]: Is all of this HTML needed for the test, as the initialise for the multi-select is only looking for the presence of the `tna-multi-select-search` class on the page?

**Review by TomJKing — COMMENTED** [review]
Looks good. Just some minor comments. Thanks

File: npm/test/index.test.ts (hunk 9 above)
Comment (vimleshtna) [reply npm/test/index.test.ts:127]: Yeah, we need it. It checks all the child elements while initialising the `tna-multi-select-search` element.

**Review by TomJKing — APPROVED** [review]
Looks good. Thanks
